lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -101,62 +115,128 @@ bool SessionStateEnforcer::m5g_init_session_credit(
+uint32_t SessionStateEnforcer::update_session_rules(
+    const std::string& imsi, std::unique_ptr<SessionState>& session_state,
+    SessionStateUpdateCriteria& session_uc, bool gnb_teid_get,
+    bool upf_teid_get) {
+  uint32_t upf_teid = 0;
+  SetGroupPDR rule;
+
+  // Get the latest config
+  const auto& config = session_state->get_config();
+  auto itp           = pdr_map_.equal_range(imsi);
+  // Lets take local_teid of the session if already exists.
+  // Not needed in 2nd AMF request processing though
+  if (!upf_teid_get) {
+    upf_teid = session_state->get_upf_local_teid();
+  }
   for (auto itr = itp.first; itr != itp.second; itr++) {
     // Get the PDR numbers, now  get the rules from global static rule list
-    SetGroupPDR rule;
     GlobalRuleList.get_rule(itr->second, &rule);
-    session_state.insert_pdr(&rule);
-    // Get new TEID
-    if (!l_teid) l_teid = pipelined_client_->get_current_teid();
-    rule.mutable_pdi()->set_local_f_teid(pipelined_client_->get_next_teid());
+    // Get the UE ip address
+    rule.mutable_pdi()->set_ue_ip_adr(
+        config.rat_specific_context.m5gsm_session_context()
+            .pdu_address()
+            .redirect_server_address());
+    rule.mutable_activate_flow_req()->mutable_sid()->set_id(imsi);
+    rule.mutable_deactivate_flow_req()->mutable_sid()->set_id(imsi);
+    rule.mutable_activate_flow_req()->set_ip_addr(
+        config.rat_specific_context.m5gsm_session_context()
+            .pdu_address()
+            .redirect_server_address());
+    rule.mutable_deactivate_flow_req()->set_ip_addr(
+        config.rat_specific_context.m5gsm_session_context()
+            .pdu_address()
+            .redirect_server_address());
+    switch (rule.pdi().src_interface()) {
+      case ACCESS:
+        // Get new UPF TEID
+        if (upf_teid_get) {
+          upf_teid = insert_pdr_from_access(session_state, session_uc, rule);
+        }
+        break;
+      case CORE:
+        if (gnb_teid_get) {
+          if (!insert_pdr_from_core(imsi, session_state, session_uc, rule))
+            return 0;
+        }
+        break;
+    }
   }
-  auto ip_addr = session_state.get_config()
-                     .rat_specific_context.m5gsm_session_context()
-                     .pdu_address()
-                     .redirect_server_address();
+  return upf_teid;
+}
+
+bool SessionStateEnforcer::handle_session_init_rule_updates(
+    SessionMap& session_map, const std::string& imsi,
+    std::unique_ptr<SessionState>& session_state) {
+  uint32_t upf_teid = 0;
+  bool gnb_teid_get = false;
+  bool upf_teid_get = true;
+  /* Add default rules to the IMSI set rules */
+  add_default_rules(session_state, imsi);
+
+  auto session_id = session_state->get_session_id();
+  /* Irrespective of any State of Session, release and terminate*/
+  SessionUpdate session_update =
+      SessionStore::get_default_session_update(session_map);
+  SessionStateUpdateCriteria session_uc = session_update[imsi][session_id];

```

Comment:
I think `SessionStateUpdateCriteria session_uc = session_update[imsi][session_id];` might be a no-op here. You either want a reference to the object in the map or a pointer. (I prefer pointer for extra clarity.)

Reply:
I'm a bit lost in the logic here but do we even need to fill in `SessionUpdate` here? If not let's pass in `nullptr` into `update_session_rules` so that it's crystal clear it's a no-op.

Reply:
Modified as a pointer type.